src/schema-parser/avro-schema-parser.ts

```diff
@@ -0,0 +1,347 @@
+ * the map of props contained by the record. The record will also be cached.
+ *
+ * @param avroDefinition the avro schema to be processed
+ * @param recordCache the cache of previously processed avro record types
+ * @param jsonSchema the schema for the record.
+ * @returns {Promise<Map<string, any>>}
+ */
+async function processRecordSchema(avroDefinition: AvroSchema, recordCache: {[key:string]: AsyncAPISchema}, jsonSchema: AsyncAPISchema) {
+  const propsMap = new Map();
+  for (const field of avroDefinition.fields) {
+    // If the type is a sub schema it will have been stored in the cache.
+    if (recordCache[field.type]) {
+      propsMap.set(field.name, recordCache[field.type]);
+    } else {
+      const def = await convertAvroToJsonSchema(field.type, false, recordCache);
+
+      requiredAttributesMapping(field, jsonSchema, field.default !== undefined);
+      commonAttributesMapping(field, def, false);
+      additionalAttributesMapping(field.type, field, def);
+
+      propsMap.set(field.name, def);
+      // If there is a name for the sub record cache it under the name.
+      const qualifiedFieldName = getFullyQualifiedName(field.type);
+      cacheAvroRecordDef(recordCache, qualifiedFieldName, def);
+    }
+  }
+  return propsMap;
+}
+
+/**
+ * Handles processing union avro schema types by creating a oneOf jsonSchema definition. This will mutate the passed
+ * jsonSchema and recordCache objects.
+ *
+ * @param jsonSchema the jsonSchema object that will be mutated.
+ * @param avroDefinition the avro schema to be processed
+ * @param isTopLevel is this the top level of the schema or is this a sub schema
+ * @param recordCache the cache of previously processed record types
+ * @returns {Promise<JsonSchema>} the mutated jsonSchema that was provided to the function
+ */
+async function processUnionSchema(jsonSchema: AsyncAPISchema, avroDefinition: AvroSchema, isTopLevel: boolean, recordCache: {[key:string]: AsyncAPISchema}) {
+  jsonSchema.oneOf = [];
+  let nullDef = null;
+  for (const avroDef of avroDefinition) {
+    const def = await convertAvroToJsonSchema(avroDef, isTopLevel, recordCache);
+    // avroDef can be { type: 'int', default: 1 } and this is why avroDef.type has priority here
+    const defType = avroDef.type || avroDef;
+    // To prefer non-null values in the examples skip null definition here and push it as the last element after loop
+    if (defType === 'null') {
+      nullDef = def;
+    } else {
+      jsonSchema.oneOf.push(def);
+      const qualifiedName = getFullyQualifiedName(avroDef);
+      cacheAvroRecordDef(recordCache, qualifiedName, def);
+    }
+  }
+  if (nullDef) jsonSchema.oneOf.push(nullDef);
+
+  return jsonSchema;
+}
+
+export async function avroToJsonSchema(avroDefinition: AvroSchema) {
+  validateAvroSchema(avroDefinition); // TODO: this is done in 'validate()', so I don't think we need to do it here
```

Comment:
`parse` function will only run when given schema is valid, so it's not needed :)